app/services/conversation_ai/stream_handler.rb

```diff
@@ -0,0 +1,30 @@
+class ConversationAi::StreamHandler
+  def initialize(reply)
+    @reply = reply
+  end
+
+  def handle_stream
+    proc do |chunk, _bytesize|
+      new_content = chunk.dig("choices", 0, "delta", "content")
+      finish_reason = chunk.dig("choices", 0, "finish_reason")
+
+      new_content = format_content(@reply, new_content)
+
+      if new_content.present?
+        @reply.content += new_content
+        @reply.broadcast_updated(new_content)
+      end
+
+      @reply.save! if finish_reason.present?
+    end
```

Comment:
I haven't fully wrapped my mind around this separate of concerns. The fact that you're passing in an active record model and then saving it here smells a bit off.

It does sometimes make sense to create models that are not backed by active record. However, that's typically when a model has a really isolated concern. But when this model is so tightly coupled to an activerecord model means that these should probably be methods on the Reply model itself, or maybe there is indeed a separate of concerns but you haven't gotten the boundary layer right yet. A clean boundary layer would mean not passing in an activerecord model and saving it here.

I'll keep reading and see if I can have a better proposal for you. I don't quite get what a "StreamHandle" is and what a "TokenHandle" is